framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,146 @@
+package org.checkerframework.common.accumulation;
+
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.Tree.Kind;
+import java.util.Arrays;
+import java.util.List;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.dataflow.analysis.FlowExpressions;
+import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.dataflow.analysis.TransferResult;
+import org.checkerframework.dataflow.cfg.node.MethodInvocationNode;
+import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.framework.flow.CFAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+
+/**
+ * The default transfer function for an accumulation checker.
+ *
+ * <p>Subclasses should call the {@link #accumulate(Node, TransferResult, String...)} accumulate}
+ * method to add a string to the estimate at a particular program point.
+ */
+public class AccumulationTransfer extends CFTransfer {
+
+    /** The type factory. */
+    protected final AccumulationAnnotatedTypeFactory typeFactory;
+
+    /**
+     * Build a new AccumulationTransfer for the given analysis.
+     *
+     * @param analysis the analysis
+     */
+    public AccumulationTransfer(CFAnalysis analysis) {
+        super(analysis);
+        typeFactory = (AccumulationAnnotatedTypeFactory) analysis.getTypeFactory();
+    }
+
+    /**
+     * Updates the estimate of how many things {@code node} has accumulated.
+     *
+     * <p>If the node is an invocation of a method that returns its receiver, then its receiver's
+     * type will also be updated. In a chain of method calls, this process will continue as long as
+     * each receiver is itself a receiver-returning method invocation.
+     *
+     * <p>For example, suppose {@code node} is the expression {@code a.b().c()}, the new value
+     * (added by the accumulation analysis because of the {@code .c()} call) is "foo", and b and c
+     * return their receiver. Then all of the expressions {@code a.b().c()}, {@code a.b()}, and
+     * {@code a} would have their estimates updated to include "foo". Note that due to what kind of
+     * values can be held in the store, this information is lost outside the method chain. That is,
+     * the returns-receiver propagated information is lost outside the expression in which the
+     * returns-receiver method invocations are nested.
+     *
+     * <p>As a concrete example, consider the Called Methods accumulation checker: if {@code build}
+     * requires a, b, and c to be called, then {@code foo.a().b().c().build();} will typecheck (they
+     * are in one fluent method chain), but {@code foo.a().b().c(); foo.build();} will not -- the
```

Comment:
The test case exists, in `SimpleFluent.java`. It passes, because the Test Accumulation called-methods checker includes the other rules that allows the code to typecheck. I'll clarify this documentation to explicitly mention the need for another rule from the user if they want to handle this kind of case.

Reply:
The test case I had in mind contained a non-`@This`-returning method in the middle of a chain of `@This`-returning methods.  That is, there appear `@This`-returning methods both before and after it.  I added such a test case, and it fails.  It is poorly named `m8` and you might want to rename it.